clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3675,14 +4062,260 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     removeFieldMemberExpr(FD, ArrayType);
     return true;
   }
+  using SyclKernelFieldHandler::enterArray;
+  using SyclKernelFieldHandler::enterStruct;
+  using SyclKernelFieldHandler::handleNonDecompStruct;
+  using SyclKernelFieldHandler::handlePointerType;
+  using SyclKernelFieldHandler::handleScalarType;
+  using SyclKernelFieldHandler::handleSyclSpecialType;
+  using SyclKernelFieldHandler::handleUnionType;
+  using SyclKernelFieldHandler::leaveArray;
+  using SyclKernelFieldHandler::leaveStruct;
+};
+
+class FreeFunctionKernelBodyCreator : public SyclKernelFieldHandler {
+  SyclKernelDeclCreator &DeclCreator;
+  llvm::SmallVector<Stmt *, 16> BodyStmts;
+  FunctionDecl *FreeFunc;
+  SourceLocation KernelCallerSrcLoc; // KernelCallerFunc source location.
+  llvm::SmallVector<Expr *, 8> ArgExprs;
+
+  // Creates a DeclRefExpr to the ParmVar that represents the current field.
```

Comment:
Please update the comments since the ones from the original BodyCreator don't make sense here.  Change current field to current free function parameter.